### PR TITLE
Add support for new libbpf APIs

### LINKED ISSUE #1285: Add new libbpf APIs
### Describe the feature you'd like supported

Add support for recently added libbpf APIs:

```
LIBBPF_API int libbpf_prog_type_by_name(const char *name, enum bpf_prog_type *prog_type, enum bpf_attach_type *expected_attach_type);
LIBBPF_API int libbpf_attach_type_by_name(const char *name, enum bpf_attach_type *attach_type);
LIBBPF_API const char *libbpf_bpf_attach_type_str(enum bpf_attach_type t);
LIBBPF_API const char *libbpf_bpf_link_type_str(enum bpf_link_type t);
LIBBPF_API const char *libbpf_bpf_map_type_str(enum bpf_map_type t);
LIBBPF_API const char *libbpf_bpf_prog_type_str(enum bpf_prog_type t);
```

### Proposed solution

Add support for these APIs

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dthaler**
## Description

* Add `libbpf_bpf_{attach,link,map,prog}_type_str()`.  These replace our previous `ebpf_get_{attach,program}_type_name()` APIs.
* Add `libbpf_attach_type_by_name()`
* Add `libbpf_strerror()`
* Add `bpf_program__insn_cnt()`
* Update map type and link type names to match Linux conventions, for apps or scripts that depend on literal string values
* Change uses of the deprecated `bpf_create_map()` to use the non-deprecated `bpf_map_create()` instead
* Add a couple of functions to the doxygen docs that were exported by missing from docs
* Use `__declspec(deprecapted)` to mark deprecated APIs as deprecated so they generate compiler warnings if used
* Update test_helper.cpp to clear the cache so that object IDs are the same for a test case regardless of whether another test was run first.  Previously running a test case by itself could see different IDs than running all cases, since the IDs vary by whether the cache was populated (which consumes an ID counter)

This PR updates ebpf-for-windows to use the latest libbpf.h file from libbpf, which removes some deprecated APIs that we still support, to ease porting apps from Linux to Windows.  Hence the prototypes for those APIs have been moved to
* include\bpf\bpf_legacy.h
* include\bpf\libbpf_legacy.h
 
Fixes #1285 

Not done in this PR, but left for a future PR (covered by issue #1295):
* Mark `bpf_prog_load_deprecated()` with the deprecated declspec (this is the only `@deprecated` API without this marking, since there are lots of callers to clean up) and clean up callers
* Clean up the deprecated use of `bpf_object__for_each_safe()` in libs/ebpfnetsh/programs.cpp

## Testing

This PR adds new tests and updates existing tests

## Documentation

This PR includes API doc updates

### CODE REVIEW

File: libs/api/libbpf_program.cpp [hunk 1]
```diff
@@ -434,17 +434,37 @@ bpf_prog_get_next_id(uint32_t start_id, uint32_t* next_id)
...
+    if (name == nullptr || attach_type == nullptr) {
+        return libbpf_err(-EINVAL);
+    }
+
+    const ebpf_attach_type_t* ebpf_attach_type = get_attach_type_windows(name);
+    if (ebpf_attach_type == &EBPF_ATTACH_TYPE_UNSPECIFIED) {
+        return libbpf_err(-ESRCH);
+    }
+
+    *attach_type = get_bpf_attach_type(ebpf_attach_type);
```
Comment (saxena-anurag) [inline libs/api/libbpf_program.cpp:467]: Same as earlier, we can instead have one call to `get_bpf_program_and_attach_type()`

File: libs/ebpfnetsh/programs.cpp [hunk 2]
```diff
@@ -716,7 +717,7 @@ handle_ebpf_show_programs(
                     execution_type_name = "NATIVE";
                     break;
                 }
-                const char* program_type_name = ebpf_get_program_type_name(&info.type_uuid);
+                const char* program_type_name = libbpf_bpf_prog_type_str(info.type);
```
Comment (saxena-anurag) [inline libs/ebpfnetsh/programs.cpp:720]: Minor: same as earlier -- I think we should use GUID here.

File: libs/api/bpf_syscall.cpp [hunk 3]
```diff
@@ -23,9 +23,11 @@ bpf(int cmd, union bpf_attr* attr, unsigned int size)
     case BPF_LINK_GET_NEXT_ID:
         CHECK_SIZE(next_id);
         return bpf_link_get_next_id(attr->start_id, &attr->next_id);
-    case BPF_MAP_CREATE:
+    case BPF_MAP_CREATE: {
         CHECK_SIZE(map_flags);
-        return bpf_create_map(attr->map_type, attr->key_size, attr->value_size, attr->max_entries, attr->map_flags);
+        struct bpf_map_create_opts opts = {.map_flags = attr->map_flags};
```
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/api/bpf_syscall.cpp:28]: probably, but the old code didn't.

[the remark this answers is not in the document]
Comment (dthaler) [reply libs/api/bpf_syscall.cpp:28]: Currently bpf_attr has no inner_map_fd field.   We can add one but that would be a separate PR.

File: libs/api/libbpf_program.cpp [hunk 4]
```diff
@@ -434,17 +434,37 @@ bpf_prog_get_next_id(uint32_t start_id, uint32_t* next_id)
...
+    ebpf_result_t result = ebpf_get_program_type_by_name(name, &ebpf_program_type, &ebpf_attach_type);
     if (result != EBPF_SUCCESS) {
         ebpf_assert(result == EBPF_KEY_NOT_FOUND);
-        errno = ESRCH;
-        return -1;
+        return libbpf_err(-ESRCH);
     }
 
+    *prog_type = get_bpf_program_type(&ebpf_program_type);
+    *expected_attach_type = get_bpf_attach_type(&ebpf_attach_type);
```
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/api/libbpf_program.cpp:450]: I don't see any API called get_bpf_program_and_attach_type in the code base.

File: libs/api/libbpf_program.cpp (hunk 1 above)
Comment (dthaler) [reply libs/api/libbpf_program.cpp:467]: I don't see any API called get_bpf_program_and_attach_type in the code base.

File: libs/api/bpf_syscall.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (saxena-anurag) [reply libs/api/bpf_syscall.cpp:28]: Yeah it looks like the header file under include\linux does not contain inner_map_fd (I think I had looked at the definition in the external path).

File: libs/api/libbpf_program.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (saxena-anurag) [reply libs/api/libbpf_program.cpp:450]: This API was removed as part of this PR itself :)

https://github.com/microsoft/ebpf-for-windows/blob/1160f7914e43ebac5e3619d662754eef1af02fb8/libs/api_common/windows_platform_common.cpp#L322

File: libs/ebpfnetsh/links.cpp [hunk 5]
```diff
@@ -47,7 +48,7 @@ handle_ebpf_show_links(
         struct bpf_link_info info;
         uint32_t info_size = (uint32_t)sizeof(info);
         if (bpf_obj_get_info_by_fd(link_fd, &info, &info_size) == 0) {
-            const char* attach_type_name = ebpf_get_attach_type_name(&info.attach_type_uuid);
+            const char* attach_type_name = libbpf_bpf_attach_type_str(info.attach_type);
```
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/ebpfnetsh/links.cpp]: Done

File: libs/ebpfnetsh/programs.cpp (hunk 2 above)
Comment (dthaler) [reply libs/ebpfnetsh/programs.cpp]: Done

File: libs/api/libbpf_program.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/api/libbpf_program.cpp]: Thanks, updated

File: libs/api/libbpf_program.cpp (hunk 1 above)
Comment (dthaler) [reply libs/api/libbpf_program.cpp]: Updated

**Review by saxena-anurag — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**